Support full-line comments in fparser2

arporter opened the pull request:
This PR is for work done under #68.

arporter commented:
I propose to leave the bug when generating fixed-format comments since that's not new and this PR is big enough already. Ready for first review by @rupertford.

rupertford commented on src/fparser/api.py:
```diff
@@ -224,7 +226,8 @@ def parse(input, isfree=None, isstrict=None, include_dirs=None,
     from .parsefortran import FortranParser
-    reader = get_reader(input, isfree, isstrict, include_dirs, source_only)
+    reader = get_reader(input, isfree, isstrict, include_dirs, source_only,
+                        ignore_comments=ignore_comments)
```
Do we want to default to ignoring comments, now that we support them?

arporter replied:
I think so given that that was effectively the behaviour before.

rupertford commented on src/fparser/Fortran2003.py:
```diff
@@ -156,20 +158,23 @@ class Base(ComparableMixin):
         :param parent_cls: the parent class of this object
         :type parent_cls: :py:type:`type`
         """
+        from . import readfortran
```
I thought relative imports were bad?

arporter replied:
I'm unsure :-) Anyway, I've made it explicit.

rupertford commented on src/fparser/Fortran2003.py:
```diff
@@ -268,38 +287,87 @@ class BlockBase(Base):
+        :param type endcls: The class marking the end of the block
+        :param reader: Content to check for match
+        :type reader: str or instance of :py:class:`FortranReaderBase`
+        :param bool match_labels: TBD
```
Lots of TBD entries here. Please either justify why or update.

arporter replied:
I'd adopted the principle of just documenting what I understood and leaving anything I didn't (on the basis that there was no interface doc at all before).

rupertford commented on src/fparser/Fortran2003.py:
```diff
@@ -268,38 +287,87 @@ class BlockBase(Base):
         content = []
+
         if startcls is not None:
+            # Deal with any preceding comments
```
What about comments after the block? I can see that these might be picked up by the next block but what about the last block in a program. Is it guaranteed that we never get into a case where comments are not dealt with?

arporter replied:
I've added to an existing test and comments after an 'end program' at least are picked up. However, as you comment below, we really need tests for this for all types of block.

arporter commented on src/fparser/api.py:
```diff
@@ -135,10 +135,12 @@ def get_reader(input, isfree=None, isstrict=None, include_dirs=None,
             return parse(c_input, isfree, isstrict, include_dirs)
         reader = FortranFileReader(input, include_dirs=include_dirs,
-                                   source_only=source_only)
+                                   source_only=source_only,
```
Good point. Done.

arporter commented on src/fparser/readfortran.py:
```diff
@@ -217,7 +221,9 @@ class Line(object):
         self.line = line.strip()
-        assert self.line, repr((line, linenospan, label))
+        if not self.line:
+            raise Exception("Got empty line: '{0}'. linenospan={1}, "
```
Good point. Done.

arporter commented on src/fparser/readfortran.py:
```diff
@@ -578,16 +605,16 @@ def get_single_line(self, ignore_empty=False, ignore_comments=False):
-
-        if ignore_comments and _is_fix_comment(line, isstrict=self.isstrict):
-            return self.get_single_line(ignore_empty, ignore_comments)
+        if ignore_comments and (self.isfix or self.isf77):
```
I had to add the isfix and isf77 checks in order that I can safely call the _is_fix_comment() routine.
I've added a comment to explain this if case.

arporter commented on src/fparser/readfortran.py:
```diff
@@ -666,41 +702,47 @@ def next(self, ignore_comments = False):
         except StopIteration:
             raise
-        except:
+        except Exception as err:
```
I agree. In fact it has already caused me significant grief in debugging the issues I had after merging with trunk. However, I don't know what specific things we should be catching at this point and this feels like a separate issue.

arporter commented on src/fparser/readfortran.py:
```diff
@@ -666,41 +702,47 @@ def next(self, ignore_comments = False):
             message = ''.join(('Traceback\n',
                               ''.join( traceback.format_stack())))
             logging.getLogger(__name__).debug(message)
+            logging.getLogger(__name__).debug(str(err))
```
No, an enhancement - you now get the text of the original exception logged.

arporter commented on src/fparser/readfortran.py:
```diff
@@ -881,44 +897,43 @@ def handle_cf2py_start(self, line):
-        -----
-        In-line comments are separated from line and put back to fifo
-        sequence where it will be processed as comment line.
+        Any in-line comment is extracted from line. If
```
Done.

arporter commented on src/fparser/readfortran.py:
```diff
@@ -944,6 +959,8 @@ def handle_inline_comment(self, line, lineno, quotechar=None):
 
     def handle_multilines(self, line, startlineno, mlstr):
+        '''
+        '''
```
Oops :-) I think I started doing that when I thought I was going to have to modify a method but it seems that in this case I never actually did...

arporter commented on src/fparser/readfortran.py:
```diff
@@ -1081,23 +1101,28 @@ def get_source_item(self):
+            # get_next_line does not actually consume lines - they are put
+            # into the FILO buffer as well as being returned
+            while _is_fix_cont(self.get_next_line(ignore_empty=True,
+                                                  ignore_comments=True)):
```
Surprisingly, no, this is correct. I have elaborated further in the comment but basically this piece of code just deals with constructing a single line from multiple continued lines. Once that's done, parsing continues by pulling items out of the FIFO and it is then that we get any comments.

arporter commented:
I've now updated the documentation and added the beginnings of a developers' guide. I probably need to add more to this before giving it back for review.

arporter commented:
Ready for 2nd review...
